47 run linting formatting and testing in parallel

Sohn123 opened the pull request:
I introduced unity builds for linting. Inspired by [this article](https://www.methodpark.de/blog/how-to-speed-up-clang-tidy-with-unity-builds/), unity builds make linting faster by parsing the headers less often. It was necessary to disable one check, because the generated unity build file makes suspicious includes. But I think the tradeoff is worth it, because linting now only takes 1 minute in comparison to over five.

rgwohlbold commented on .github/workflows/check.yml:
```diff
@@ -1,22 +1,33 @@
         run: make format-check
+  linting:
+    runs-on: ubuntu-latest
+    container: ghcr.io/wimove-oss/wimove-testimage:latest
```
We should put the Dockerfile for this image into the repository and ideally rebuild and push it periodically

rgwohlbold replied:
Also, we should rethink the naming: Maybe something like `wimoved-ci-env` or `wimoved-test-env`?

rgwohlbold replied:
Maybe a new directory in the `scripts/` directory with the name of the image (e.g. `wimoved-ci-env`) as its name? We need to tidy this up more in the future

rgwohlbold commented on CMakeLists.txt:
```diff
@@ -30,6 +30,7 @@ add_custom_target(test DEPENDS wimoved_tests COMMAND ./wimoved_tests)
 add_custom_target(format COMMAND clang-format -i ${SRC_FILES} ${TEST_FILES})
 add_custom_target(format-check COMMAND clang-format --dry-run --Werror ${SRC_FILES} ${TEST_FILES})
 add_custom_target(lint COMMAND run-clang-tidy -j8 -quiet ${SRC_FILES} ${TEST_FILES})
+add_custom_target(unity-lint COMMAND run-clang-tidy -quiet -p . CMakeFiles/wimoved.dir/ CMakeFiles/wimoved_tests.dir/)
```
When unity builds are enabled, this lint command seems to run much faster than the regular lint. Could we define the `lint` target conditionally such that `make precommit` calls the fast lint target when unity builds are enabled?

rgwohlbold reviewed (changes requested):
I tested it on my machine, I love the speedup!
We need some additional changes in `.clang-tidy` so the linting keeps working: Since all source files are included, we need to adjust the `HeaderRegex` to something like

```yml
HeaderFilterRegex: '(src|test)/.*\.(h|cpp)'
```

Please also take a look at my other comments

Sohn123 commented:
> ```yaml
> HeaderFilterRegex: '(src|test)/.*\.(h|cpp)'
> ```

Is there an easy way to test, that the new regex is working?

rgwohlbold commented:
> > ```yaml
> > HeaderFilterRegex: '(src|test)/.*\.(h|cpp)'
> > ```
> 
> Is there an easy way to test, that the new regex is working?

Make some change that creates a linting problem (e.g. rename a variable `name` to `someName`) and run `unity-lint`

Sohn123 commented:
@rgwohlbold the linter is unhappy now, but I haven't changed anything regarding the code. Do you think this is a false positive or do we need to fix this?

Sohn123 commented:
> > @rgwohlbold the linter is unhappy now, but I haven't changed anything regarding the code. Do you think this is a false positive or do we need to fix this?
> 
> Since the linter is now able to analyze all source files at the same time, it can perform better analysis. Before, it could not really detect which functions can throw exceptions when linting `main.cpp`, but that should be possible now. Therefore, I don't think that this is a false positive.
> 
> We have configured a termination handler so we log the exception properly, so I don't think that this is a problem here. I would feel bad about globally blacklisting this since the rule does a couple of other useful things (https://clang.llvm.org/extra/clang-tidy/checks/bugprone/exception-escape.html). Maybe add a `NOLINTNEXTLINE` thing?

I added the NOLINT directive. But now we can't get any warning

> > @rgwohlbold the linter is unhappy now, but I haven't changed anything regarding the code. Do you think this is a false positive or do we need to fix this?
> 
> Since the linter is now able to analyze all source files at the same time, it can perform better analysis. Before, it could not really detect which functions can throw exceptions when linting `main.cpp`, but that should be possible now. Therefore, I don't think that this is a false positive.
> 
> We have configured a termination handler so we log the exception properly, so I don't think that this is a problem here. I would feel bad about globally blacklisting this since the rule does a couple of other useful things (https://clang.llvm.org/extra/clang-tidy/checks/bugprone/exception-escape.html). Maybe add a `NOLINTNEXTLINE` thing?

I added the NOLINT directive, but now we can't get linting errors for this line in general. But I haven't found a way to only suppress a specific warning for this line.

rgwohlbold commented:
It should be possible via `// NOLINTNEXTLINE(bugprone-exception-escape)`, see `vendor/loginit.h`

Sohn123 commented:
Ok, thank you! I'm happy now. So maybe take another look and then we can :shipit:

rgwohlbold commented on wimoved-ci-env/Dockerfile:
```diff
@@ -0,0 +1,3 @@
+#LABEL org.opencontainers.image.source https://github.com/wimove-oss/wimoved
```
Is there a purpose to this comment? If not, we should remove it

rgwohlbold reviewed (changes requested):
Looks good! Just two small things and we are ready to go :rocket:

Sohn123 commented:
> Also, I find the `wimove-ci-env` directory a bit annoying in the root directory. Could we move it in the `scripts/` directory?

I didn't want to move it in the scripts directory because then it would be part of the build context for the build environment. I think we have to restructure the scripts directory and the ci image directory in general. I can start with this afternoon but it is out of scope for this PR

rgwohlbold reviewed (approved):
Looks good now :shipit:
